guides/hack/06-classes/16-constructors.md

```diff
@@ -11,28 +11,49 @@ being incremented.  Note that a constructor may call any *private* method in its
-programming shortcut by having the implementation declare and initialize such properties.  However, this approach violates data hiding
-by admitting publicly in the constructor's calling interface the name and type of a private data member.
+@@ constructors-examples/duplication.hack @@
 
-Can we use this approach with the `Point` class?
-No, not as we've designed it.  Our private properties have type `float`, so any arithmetic coordinate value can be
-represented, yet we've declared the parameters to the constructor to have type `num`, so either integer or floating-point values can
-be passed in.  Specifically, the type of the private members is *not* the same as their corresponding parameters!  Using the following
-short-hand notation:
+The class properties are essentially repeated multiple times: at declaration, in
+the constructor parameters and in the assignment. This can be quite cumbersome.
 
-```Hack
-public function __construct(private num $x = 0, private num $y = 0) { ... }
-```
+With *constructor parameter promotion*, all that repetitive boilerplate is
+removed.
 
-this results in both private properties having type `num` instead of `float`, and that we don't want!
+@@ constructors-examples/promotion.hack @@
 
-While it is often the case that there is a one-for-one correspondence between a constructor's parameters and the class's private
-properties, this need not be so.  In fact, that's one reason to use data hiding.  For example, an alternate way of representing a
-point is to use Polar Coordinates; that is, using an angle and a distance from the origin.  **Don't advertise/promise the internal
-representation of an object in its constructor argument list!**
+All you do is put a visibility modifier in front of the constructor parameter
+and everything else in the previous example is done automatically, including the
+actual creation of the property.
+
+### Rules
+
+* A modifier of `private`, `protected` or `public` must precede the parameter
+  declaration in the constructor.
+* Other, non-class-property parameters may also exist in the constructor, as
+  normal.
+* Type annotations must go between the modifier and the parameter's name.
+* The parameters can have a default value.
+* Other code in the constructor is run **after** the parameter promotion
+  assignment.
+
+@@ constructors-examples/promotion-rules.hack @@
+
+### Design considerations
+
+Can we use constructor parameter promotion with the `Point` class? No, not as
+we've designed it. Our private properties have type `float`, so any arithmetic
+coordinate value can be represented, yet we've declared the parameters to the
+constructor to have type `num`, so either integer or floating-point values can
+be passed in.
+
+As this example illustrates, it is often valuable to keep the internal data
+representation separate from the public-facing constructor, which is impossible
+with promoted constructor parameters.
```

Comment:
I know this is copy-paste from the previous version, but I'd like this removed: it does not affect callers at all.

This is as true as saying that "it's impossible to directly return an awaitable without awaiting it if you have `async` on your function declaration header"

Both this and the previous paragraph should probalby be replaced with a generic warning about public/protected properties in general, but there is no additional harm from using paramter promotion.

Reply:
Yeah you're probably right, I should make that section even shorter or maybe just remove it.

I guess the main argument is not as much that promoted params are bad, but that they could lead people towards bad design decisions (making your constructor take arguments equal to your private properties, instead of some more reasonable public API).